In [230]:
def read_data(use_input = False):
    data = ""
    fname = "day2.input" if use_input else "day2.example"
    with open(fname, "r") as f:
        data = f.read()

    # split into reports
    reports = data.split("\n")
    return reports

def isSafe(line:str):
    report = [
        int(n) for n in line.split(" ")
    ]
    sign = (report[0] - report[1]) > 0

    for i in range(len(report)-1):
        if (report[i] - report[i+1] > 0) != sign:
            return False, i
        if not(0< abs(report[i] - report[i+1]) <4):
            return False, i
        
    return True, None
            



In [231]:

safe_count = 0
for line in read_data() :
    safety = isSafe(line)
    if safety[0]: 
        safe_count += 1

print(f"{safe_count=}")


safe_count=2


In [234]:
import copy

def bruteForceLine(line):
    line2 = [ int(n) for  n in line.split(" ") ]
    res = []
    for i in range(len(line2)):
        line_tmp = copy.copy(line2)
        line_tmp.pop(i)
        res.append(
            " ".join([ str(n) for n in line_tmp ]).strip()
        )

    return res

def removeElem(line, indexElem):
    if not( 0<=indexElem<len(line)):
        return "1 99" # cause always UNSAFE
    line_tmp = copy.copy(line)
    line_tmp = [ int(n) for  n in line_tmp.split(" ") ]
    line_tmp.pop(indexElem)
    return " ".join([ str(n) for n in line_tmp ]).strip()

safe_count = 0
for line in read_data(True) :
    safety = isSafe(line)
    
    if safety[0] :
        safe_count += 1
    else:
        attempt1 = isSafe(removeElem(line, safety[1]-1))[0]
        attempt2 = isSafe(removeElem(line, safety[1]))[0]
        attempt3 = isSafe(removeElem(line, safety[1]+1))[0]
        if any((attempt1, attempt2, attempt3)):
            safe_count += 1



print(f"\n{safe_count=}")


safe_count=364


['1', '3', '2', '4', '5'] [2, -1, 2, 1] [-3, 3, -1]

sia canvando il '3' che cavando il '2', rendo il report sicuro
nell esempio "suggeriscono" di cavare il secondo
probabilmente perché è un monte tra due valli (il report non è valido perche non è crescente)
in questo caso il problema è che 


337 < right_answer < 388
 right_answer = 364